### PR TITLE
ASRC-131 BUN Lab Result

### PR DESCRIPTION
**Pull request by dgtombs**
Though I did add the BUN Lab Result on this branch, the big change was introducing the risk model objects. There is no longer direct navigability from a Specialty to Variables, instead it is Specialty -> Model(s) -> Term(s) -> Variables.

### CODE REVIEW

File: srcalc/src/main/java/gov/va/med/srcalc/db/SpecialtyDao.java [hunk 1]
```diff
@@ -31,31 +31,21 @@ protected Session getCurrentSession()
...
-            // values in the variable list if the database is misconfigured.
-            // Detect this early to avoid mystery NPEs later.
-            if (var == null)
-            {
-                throw new ConfigurationException(String.format(
-                        "Missing Variable for specialty %s. Check the order column.",
-                        name));
-            }
-
             try
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/db/SpecialtyDao.java]: Are models non-nullable?

File: srcalc/src/main/java/gov/va/med/srcalc/domain/Calculation.java [hunk 2]
```diff
@@ -87,13 +88,51 @@ public void setSpecialty(final Specialty specialty)
                     "Cannot return list of variables because no specialty has been set.");
         }
         
-        return fSpecialty.getVariables();
+        return Collections.unmodifiableSet(fSpecialty.getModelVariables());
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/Calculation.java]: Why all of the changes from lists to sets? Is this to deal with larger input data?

File: srcalc/src/main/java/gov/va/med/srcalc/domain/model/ConstantTerm.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        // Performance optimization.
+        if (o == this)
+        {
+            return true;
+        }
+        
+        if (o instanceof ConstantTerm)
+        {
+            final ConstantTerm other = (ConstantTerm)o;
+            return baseEquals(other);
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/model/ConstantTerm.java]: You could cast "o" as a method parameter instead of storing it in other, but it's a preference thing.

File: srcalc/src/main/java/gov/va/med/srcalc/domain/model/ModelTerm.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    @Transient
+    public abstract Set<Variable> getRequiredVariables();
+    
+    /**
+     * Base equals() functionality that simply verifies equality of the coefficient.
+     * @return true if the coefficients are equal, false otherwise
+     */
+    protected boolean baseEquals(final ModelTerm other)
+    {
+        return new Float(this.getCoefficient()).equals(other.getCoefficient());
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/model/ModelTerm.java]: Is there any chance for other to be null?

File: srcalc/src/main/java/gov/va/med/srcalc/domain/variable/PopulatedVariableGroup.java [hunk 5]
```diff
@@ -20,7 +20,7 @@
      * @throws IllegalArgumentException if the given list is empty or if it
      * contains Variables with different groups
      */
-    public PopulatedVariableGroup(final List<Variable> variables)
+    public PopulatedVariableGroup(final List<? extends Variable> variables)
```
Comment (jamesbigmac) [inline srcalc/src/main/java/gov/va/med/srcalc/domain/variable/PopulatedVariableGroup.java]: Why the change here? Were you specifically trying to exclude any Variable types? I'm a bit fuzzy on the reasoning for "? extends" because I've never actually seen it used in code.

File: srcalc/src/main/java/gov/va/med/srcalc/db/SpecialtyDao.java (hunk 1 above)
Comment (dgtombs) [inline srcalc/src/main/java/gov/va/med/srcalc/db/SpecialtyDao.java]: I updated the PR to ensure the sets do not contain null.
